typegate/src/typegate/artifacts/shared.ts

```diff
@@ -0,0 +1,169 @@
+// Copyright Metatype OÜ, licensed under the Elastic License 2.0.
+// SPDX-License-Identifier: Elastic-2.0
+
+import { createHash } from "node:crypto";
+import { S3, S3ClientConfig } from "aws-sdk/client-s3";
+import {
+  deinitUploadUrlStore,
+  getLocalPath,
+  initUploadUrlStore,
+  STORE_DIR,
+  STORE_TEMP_DIR,
+  UploadUrlStore,
+} from "./local.ts";
+import { ArtifactStore } from "./mod.ts";
+import { HashTransformStream } from "../../utils/hash.ts";
+import { resolve } from "std/path/resolve.ts";
+
+export class SharedArtifactStore extends ArtifactStore {
+  #uploadUrls: UploadUrlStore;
+  #s3!: S3;
+  #s3Bucket: string;
+
+  static init(s3Config: S3ClientConfig, s3Bucket: string) {
+    return new SharedArtifactStore(s3Config, s3Bucket);
+  }
+
+  constructor(s3Config: S3ClientConfig, s3Bucket: string) {
+    super();
+    this.#s3 = new S3(s3Config);
+    this.#s3Bucket = s3Bucket;
+    this.#uploadUrls = initUploadUrlStore();
+  }
+
+  override async persist(stream: ReadableStream<any>): Promise<string> {
+    const tmpFile = await Deno.makeTempFile({
+      dir: STORE_TEMP_DIR,
+    });
+    const file = await Deno.open(tmpFile, { write: true, truncate: true });
+    const hasher = createHash("sha256");
+    await stream
+      .pipeThrough(new HashTransformStream(hasher))
+      .pipeTo(file.writable);
+
+    const hash = hasher.digest("hex");
+    await this.#s3.putObject({
+      Bucket: this.#s3Bucket,
+      Key: hash,
+      Body: stream,
```

Comment:
The stream would have been already consumed by this time...

Reply:
Yeah, right. I will read from the tmp file...

Reply:
Or directly stream it to S3, with a temporary key then [rename (Copy+Delete)](https://docs.aws.amazon.com/AmazonS3/latest/userguide/copy-object.html) after we get the hash...

Reply:
Done.